xdsl/irdl/constraints.py

```diff
@@ -1,17 +1,29 @@
 from __future__ import annotations
 
 from abc import ABC, abstractmethod
-from collections.abc import Sequence
+from collections.abc import Callable, Sequence, Set
 from dataclasses import dataclass, field
 from inspect import isclass
-from typing import Generic, TypeAlias, TypeVar
+from typing import Generic, TypeAlias, TypeVar, cast
 
 from typing_extensions import assert_never
 
 from xdsl.ir import Attribute, AttributeCovT, ParametrizedAttribute
 from xdsl.utils.exceptions import VerifyException
 from xdsl.utils.runtime_final import is_runtime_final
 
+_AttributeCovT = TypeVar("_AttributeCovT", bound=Attribute, covariant=True)
+
+ResolveType: TypeAlias = Attribute | Sequence[Attribute]
+
+
+def print_resolve_type(r: ResolveType | None) -> str:
+    if isinstance(r, Attribute):
+        return str(r)
+    elif isinstance(r, Sequence):
+        return str(tuple(str(x) for x in r))
+    return "None"
```

Comment:
```suggestion
def _format_resolve_type(r: ResolveType | None) -> str:
    if isinstance(r, Attribute):
        return str(r)
    elif isinstance(r, Sequence):
        return str(tuple(str(x) for x in r))
    return "None"
```

Reply:
Also a doc string would be good

Reply:
Seems like this is dead code so I've removed it